### PR TITLE
Change logic to merge RUNTIME_FUNCTIONs into function address ranges

### PR DESCRIPTION
**Pull request by dpallotti**
The current logic didn't consider the following from https://docs.microsoft.com/en-us/cpp/build/exception-handling-x64#chained-unwind-info-structures: "Chained info [...] can be used for noncontiguous code segments. By using chained info, you can reduce the size of the required unwind information, because you do not have to duplicate the unwind codes array from the primary unwind info."

It only considered the more common case "You can also use chained info to group volatile register saves. The compiler may delay saving some volatile registers until it is outside of the function entry prolog."

In other words, it didn't consider that `RUNTIME_FUNCTION`s can be chained with a primary `RUNTIME_FUNCTION` that actually belongs to another function and that came way earlier, as a means of reusing unwind information from that function. This is very rare as we only observed it for a handful of `RUNTIME_FUNCTION`s in one large complex game.

The new logic to merge a group of `RUNTIME_FUNCTION`s into a single function is that the `RUNTIME_FUNCTION`s must be adjacent (i.e., no "holes" allowed), and that they must have the same primary `RUNTIME_FUNCTION`. But we don't require the primary `RUNTIME_FUNCTION` to belong to the group, to support this new corner case.

Test:
- Existing unit tests of `LoadDebugSymbols`, that uses this function.
- With unit tests that will come in a following change regarding loading these `UnwindRange`s as symbols.
- As part of the larger symbol loading fallback prototype.

### CODE REVIEW

File: src/ObjectUtils/CoffFile.cpp [hunk 1]
```diff
@@ -417,49 +417,130 @@ orbit_object_utils::CoffFileImpl::GetPrimaryRuntimeFunction(
...
       continue;
     }
 
+    ORBIT_CHECK(last_primary_address != 0);
+    ORBIT_CHECK(largest_primary_address != 0);
+
+    UnwindRange& last_range = unwind_ranges.back();
+    const uint64_t last_end_address = last_range.end;
+
+    if (end_address < start_address) {
```
Comment (ronaldfw) [inline src/ObjectUtils/CoffFile.cpp:476]: I don't find it obvious why this check is down here and not further up. Would it be wrong to check for this error case right after getting the start and end address (that is, right after line 457)? It would be different than the current code (unless there are some non-obvious implications that make the following impossible), as you may still create an unwind address range using `primary_address` and `end_address` right now, even if `end_address < start_address`. But that seems like you a using potentially bad data from an illegal runtime function?

File: src/ObjectUtils/CoffFile.cpp [hunk 2]
```diff
@@ -417,49 +417,130 @@ orbit_object_utils::CoffFileImpl::GetPrimaryRuntimeFunction(
...
-      // Extend the size of the current function.
+    if (start_address < last_end_address) {
+      return ErrorMessage{
+          absl::StrCat(kErrorMessagePrefix, "RUNTIME_FUNCTIONs not sorted or overlapping.")};
+    }
+
+    // https://docs.microsoft.com/en-us/cpp/build/exception-handling-x64#struct-unwind_info states
+    // that "the chained unwind info entry is the contents of a *previous* RUNTIME_FUNCTION entry",
+    // and we rely on that here.
+    if (primary_address > start_address) {
```
Comment (ronaldfw) [inline src/ObjectUtils/CoffFile.cpp:491]: Nit: Other checks use `<`. I would find it slightly more intuitive to read if this were also `start_address < primary_address`.

File: src/ObjectUtils/CoffFile.cpp [hunk 3]
```diff
@@ -417,49 +417,130 @@ orbit_object_utils::CoffFileImpl::GetPrimaryRuntimeFunction(
...
+  // And note that we don't require the primary RUNTIME_FUNCTION to actually be part of the current
+  // set of RUNTIME_FUNCTIONs we are merging. We observed this as a very rare but valid case, on
+  // which more information can be found below.
   std::vector<UnwindRange> unwind_ranges;
-  // For a function, the "primary" RUNTIME_FUNCTION and the ones with UNW_FLAG_CHAININFO set (if
-  // any) form a tree rooted on the "primary" RUNTIME_FUNCTION. We merge all such RUNTIME_FUNCTIONs
-  // from the same tree into a single UnwindRange.
+  uint64_t last_primary_address{};
+  uint64_t largest_primary_address{};
+  uint64_t num_chained_not_with_largest_primary = 0;
```
Comment (ronaldfw) [inline src/ObjectUtils/CoffFile.cpp:446]: Reading `num_chained_not_with_largest_primary`, I don't really have an idea what this is supposed to be. Can you think of a better name or add a comment?

File: src/ObjectUtils/CoffFile.cpp [hunk 4]
```diff
@@ -417,49 +417,130 @@ orbit_object_utils::CoffFileImpl::GetPrimaryRuntimeFunction(
...
+  // primary RUNTIME_FUNCTION. However, we never observed this case.
+  //
+  // And note that we don't require the primary RUNTIME_FUNCTION to actually be part of the current
+  // set of RUNTIME_FUNCTIONs we are merging. We observed this as a very rare but valid case, on
+  // which more information can be found below.
   std::vector<UnwindRange> unwind_ranges;
-  // For a function, the "primary" RUNTIME_FUNCTION and the ones with UNW_FLAG_CHAININFO set (if
-  // any) form a tree rooted on the "primary" RUNTIME_FUNCTION. We merge all such RUNTIME_FUNCTIONs
-  // from the same tree into a single UnwindRange.
+  uint64_t last_primary_address{};
```
Comment (ronaldfw) [inline src/ObjectUtils/CoffFile.cpp:444]: Nit: last --> previous?

File: src/ObjectUtils/CoffFile.cpp [hunk 5]
```diff
@@ -417,49 +417,130 @@ orbit_object_utils::CoffFileImpl::GetPrimaryRuntimeFunction(
...
+      ORBIT_CHECK(largest_primary_address == 0);
+      last_primary_address = primary_address;
+      largest_primary_address = primary_address;
       continue;
     }
 
+    ORBIT_CHECK(last_primary_address != 0);
+    ORBIT_CHECK(largest_primary_address != 0);
+
+    UnwindRange& last_range = unwind_ranges.back();
```
Comment (ronaldfw) [inline src/ObjectUtils/CoffFile.cpp:473]: Nit: Also here, previous instead of last seems a little more precise.

**Review by ronaldfw — APPROVED** [review]
LGTM

Just some minor nits and questions

**Review by florian-kuebler — APPROVED** [review]
Thanks. I agree with the comments from @ronaldfw, and don't have anything to add.
From my side, looks good!

File: src/ObjectUtils/CoffFile.cpp (hunk 1 above)
Comment (dpallotti) [reply src/ObjectUtils/CoffFile.cpp]: Yeah, and the same applied to the "chained RUNTIME_FUNCTION is not a previous one" case. I moved both up, should be better for the reason you mentioned.

File: src/ObjectUtils/CoffFile.cpp (hunk 3 above)
Comment (dpallotti) [reply src/ObjectUtils/CoffFile.cpp]: I'm relying on the comment in `if (primary_address < largest_primary_address) {`, which I extended slightly. I would avoid adding yet another one on this matter. However, I renamed to `number_of_runtime_functions_whose_primary_is_not_the_latest_primary`, I hope it's good enough, I doubt I can do better without going over 100 characters :D

File: src/ObjectUtils/CoffFile.cpp (hunk 1 above)
Comment (ronaldfw) [reply src/ObjectUtils/CoffFile.cpp]: Ok, looks good

File: src/ObjectUtils/CoffFile.cpp (hunk 3 above)
Comment (ronaldfw) [reply src/ObjectUtils/CoffFile.cpp]: Sounds good, it's long, but it's clear

**Review by ronaldfw — APPROVED** [review]
LGTM
